feat: add snyk workflow

sarakolding opened the pull request:
snyk not part of dependencies?

MartinBernstorff commented:
Looks great Sara!

We added the comments to provide context when CI failed. Since we are a small group with relatively low turnover, I don't think they are worth the complexity (as @KennethEnevoldsen suggested), so feel free to remove them 👍 

(I know that's super annoying after you've worked hard on adding them!)

`inv snyk` was added quite rapidly as well; I'd suggest adding something like `for requirements_file in Path(".").rglob(".+req.+\.txt")` to make it robust to further additions of requirements as well 👍

sarakolding commented:
> Looks great Sara!
> 
> We added the comments to provide context when CI failed. Since we are a small group with relatively low turnover, I don't think they are worth the complexity (as @KennethEnevoldsen suggested), so feel free to remove them.
> 
> (I know that's super annoying after you've worked hard on adding them!)

That's fine, was just playing around with it a bit, and I agree --> let's delete 🐒

> `inv snyk` was added quite rapidly as well; currently, it requires hard-coding each requirements.txt path. If we add another one in the future, it won't be picked up. You can improve upon it with some regex and `rglob` if you'd like! 👍

I was going to do this but saw that we have hardcoded the names of the different requirement files in other workflows, so thought it was maybe better to keep it consistent?

sarakolding commented:
> > Looks great Sara!
> > We added the comments to provide context when CI failed. Since we are a small group with relatively low turnover, I don't think they are worth the complexity (as @KennethEnevoldsen suggested), so feel free to remove them.
> > (I know that's super annoying after you've worked hard on adding them!)
> 
> That's fine, was just playing around with it a bit, and I agree --> let's delete 🐒
> 
> > `inv snyk` was added quite rapidly as well; currently, it requires hard-coding each requirements.txt path. If we add another one in the future, it won't be picked up. You can improve upon it with some regex and `rglob` if you'd like! 👍
> 
> I was going to do this but saw that we have hardcoded the names of the different requirement files in other workflows, so thought it was maybe better to keep it consistent?

So, I deleted the context messages and Lasse and I agreed that we would keep the hardcoded file paths, to keep things consistent. Also added an error message if the existing files do not match what the code specifies. Have tested that the workflow also fails if there is a vulnerability.

KennethEnevoldsen commented on .github/workflows/snyk.yml:
```diff
@@ -0,0 +1,50 @@
+        env:
+          SNYK_TOKEN: ${{ secrets.SNYK_TOKEN }}
+        continue-on-error: false
+        run: |
+          if inv snyk; then
+            echo "Snyk check passed"
+            echo "snyk_failed=0" >> $GITHUB_OUTPUT
+          else
+            echo "Snyk check failed"
+            echo "snyk_failed=1" >> $GITHUB_OUTPUT
+            exit 1
+          fi
```
I assume this should be enough?
```suggestion
        run: |
          inv snyk
```

I would probably convert the command to inv check-packages (in case we switch out snyk).

KennethEnevoldsen replied:
inv scan-for-vulnerabities

sarakolding replied:
i'm good with the name change

KennethEnevoldsen commented on .github/workflows/snyk.yml:
```diff
@@ -0,0 +1,50 @@
+      - name: Run Snyk to check for vulnerabilities
+        id: snyk
+        env:
+          SNYK_TOKEN: ${{ secrets.SNYK_TOKEN }}
```
Do we need a token for snyk?

sarakolding replied:
yup, it needs authentication

KennethEnevoldsen commented on .github/workflows/snyk.yml:
```diff
@@ -0,0 +1,50 @@
+jobs:
+  snyk:
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: write
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        python-version: ["3.10"]
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        id: setup_python
+        with:
+          python-version: ${{ matrix.python-version}}
```
```suggestion
    steps:
      - uses: actions/checkout@v3

      - name: Set up Python
        uses: actions/setup-python@v4
        with:
          python-version: "3.10"
```

simplifying a bit more here. No need for the test grid.

KennethEnevoldsen commented on .github/workflows/snyk.yml:
```diff
@@ -0,0 +1,50 @@
+        run: npm install -g snyk
+
+      - name: Run Snyk to check for vulnerabilities
+        id: snyk
```
```suggestion
```

KennethEnevoldsen reviewed (commented):
I jumped in a did a few simplifications on the ci. Looks much simpler without the comment. Glad we made that decision.

MartinBernstorff commented on tasks.py:
```diff
@@ -524,6 +524,9 @@ def snyk(c: Context):
         "test-requirements.txt",
     ]:
+        if Path(requirements_file).exists() is False:
+            raise FileNotFoundError("One of the provided files could not be found.")
```
```suggestion
            raise FileNotFoundError(f"{requirements_file} does not exist")
```

MartinBernstorff commented:
> Lasse and I agreed that we would keep the hardcoded file paths, to keep things consistent. 

In that case, we should only have it hard-coded in one place. Or, alternatively, infer them all with `rglob`.

Makes it much easier to maintain 👍 

The snyk workflow is also failing now. Other than that, looks good to me. Good work!

KennethEnevoldsen commented on .github/workflows/snyk.yml:
```diff
@@ -0,0 +1,42 @@
+      - name: Run Snyk to check for vulnerabilities
+        env:
+          SNYK_TOKEN: ${{ secrets.SNYK_TOKEN }}
+        continue-on-error: false
```
```suggestion
```

sarakolding commented:
have added some globbing to install the requirement files, but it's not super sleek. suggestions or good enough?

MartinBernstorff commented on tasks.py:
```diff
@@ -517,22 +517,27 @@ def pr(c: Context, auto_fix: bool = True, create_pr: bool = True):
-        if Path(requirements_file).exists() is False:
-            raise FileNotFoundError(f"{requirements_file} does not exist")
+def vulnerability_scan(c: Context):
+    requirements_files = Path().parent.rglob("*requirements.txt")
+    top_level_files = [file for file in requirements_files if file.parent == Path()]
+    for requirements_file in top_level_files:
+        if requirements_file.exists() is False:
+            raise FileNotFoundError(f"{requirements_file} does not exist.")
```
Is there a need to only get top-level files? And if so, we should just `glob`, not `rglob` 👍 

```python
requirements_files = list(Path(__file__).rglob("*requirements.txt"))
for req in requirements_files:
    ...
```

MartinBernstorff replied:
No need to check the file exists anymore either; it won't be found by glob if it doesn't exist.

MartinBernstorff commented on tasks.py:
```diff
@@ -517,22 +517,27 @@ def pr(c: Context, auto_fix: bool = True, create_pr: bool = True):
 
+@task
+def install_requirements(c: Context):
+    requirements_files = Path().parent.rglob("*requirements.txt")
```
Same as above.

KennethEnevoldsen commented on .github/workflows/static_type_checks.yml:
```diff
@@ -30,7 +30,7 @@ jobs:
         run: |
           pip install invoke
-          pip install -r requirements.txt -r test-requirements.txt -r dev-requirements.txt -r gpu-requirements.txt
+          inv install-requirements
```
God idé!

KennethEnevoldsen reviewed (commented):
Ser virkelig godt ud!

MartinBernstorff commented:
Mega godt arbejde Sara! Super glad for resultatet 👍
